File: tauri/src/lib/services/settingsApi.ts
import { http, type HttpClient } from "./http";
import { serverDefaults } from "$lib/data/serverDefaults.svelte";

export interface SettingsSnapshot {
  account: string;
  model: string;
  effort: string;
  permissionMode: string;
  streaming: boolean;
  todoTools: boolean;
  browserView: boolean;
  sdkAutoUpdate: boolean;
  outputStyle: string;
  mcpDisabled: string;
  showThinking: string;
  showToolUse: string;
  showFileChange: string;
  showCompact: string;
  showWorking: string;
  showTokens: boolean;
  simpleMode: boolean;
  chatOrder: string;
  trashEnabled: boolean;
  defaultCategory: string;
  retentionDays: number;
  retentionMin: number;
  retentionMax: number;
  chatLanguage: string;
  alwaysThinking: boolean;
  autoCompact: boolean;
  remoteControl: boolean | null;
  coAuthored: boolean | null;
  sessionUpload: boolean | null;
}

export interface SettingsPatch {
  account?: string;
  model?: string;
  effort?: string;
  permission_mode?: string;
  streaming?: boolean;
  todo_tools?: boolean;
  browser_view?: boolean;
  sdk_auto_update?: boolean;
  output_style?: string;
  mcp_disabled?: string;
  show_thinking?: string;
  show_tool_use?: string;
  show_file_change?: string;
  show_compact?: string;
  show_working?: string;
  show_tokens?: boolean;
  simple_mode?: boolean;
  chat_order?: string;
  trash_enabled?: boolean;
  default_category?: string;
  retention_days?: number;
  chat_language?: string;
  always_thinking?: boolean;
  auto_compact?: boolean;
  remote_control?: boolean | null;
  co_authored?: boolean | null;
  session_upload?: boolean | null;
}

export const RETENTION_FOREVER_DAYS = 36500;

type Field = { effective?: unknown; default?: unknown; minimum?: unknown; maximum?: unknown };
type Wire = Record<string, Field | undefined>;

const boundNum = (wire: Wire, key: string, bound: "minimum" | "maximum", fallback: number): number =>
  typeof wire[key]?.[bound] === "number" ? (wire[key]![bound] as number) : fallback;

const effective = <T>(wire: Wire, key: string, kind: string, fallback: T): T => {
  const field = wire[key];
  if (!field) return fallback;
  if (typeof field.effective === kind) return field.effective as T;
  if (typeof field.default === kind) return field.default as T;
  return fallback;
};

const effectiveStr = (wire: Wire, key: string, fallback: string): string =>
  effective(wire, key, "string", fallback);

const effectiveBool = (wire: Wire, key: string, fallback: boolean): boolean =>
  effective(wire, key, "boolean", fallback);

const effectiveFlag = (wire: Wire, key: string): boolean | null =>
  typeof wire[key]?.effective === "boolean" ? (wire[key]!.effective as boolean) : null;

const effectiveNum = (wire: Wire, key: string, fallback: number): number =>
  effective(wire, key, "number", fallback);

const parse = (wire: Wire): SettingsSnapshot => ({
  account: effectiveStr(wire, "account", ""),
  model: effectiveStr(wire, "model", "opus"),
  effort: effectiveStr(wire, "effort", "xhigh"),
  permissionMode: effectiveStr(wire, "permission_mode", "bypassPermissions"),
  streaming: effectiveBool(wire, "streaming", true),
  todoTools: effectiveBool(wire, "todo_tools", false),
  browserView: effectiveBool(wire, "browser_view", false),
  sdkAutoUpdate: effectiveBool(wire, "sdk_auto_update", true),
  outputStyle: effectiveStr(wire, "output_style", "default"),
  mcpDisabled: effectiveStr(wire, "mcp_disabled", ""),
  showThinking: effectiveStr(wire, "show_thinking", "full"),
  showToolUse: effectiveStr(wire, "show_tool_use", "label"),
  showFileChange: effectiveStr(wire, "show_file_change", "full"),
  showCompact: effectiveStr(wire, "show_compact", "full"),
  showWorking: effectiveStr(wire, "show_working", "label"),
  showTokens: effectiveBool(wire, "show_tokens", false),
  simpleMode: effectiveBool(wire, "simple_mode", false),
  chatOrder: effectiveStr(wire, "chat_order", "auto"),
  trashEnabled: effectiveBool(wire, "trash_enabled", false),
  defaultCategory: effectiveStr(wire, "default_category", ""),
  retentionDays: effectiveNum(wire, "retention_days", 30),
  retentionMin: boundNum(wire, "retention_days", "minimum", 1),
  retentionMax: boundNum(wire, "retention_days", "maximum", Number.MAX_SAFE_INTEGER),
  chatLanguage: effectiveStr(wire, "chat_language", ""),
  alwaysThinking: effectiveBool(wire, "always_thinking", false),
  autoCompact: effectiveBool(wire, "auto_compact", true),
  remoteControl: effectiveFlag(wire, "remote_control"),
  coAuthored: effectiveFlag(wire, "co_authored"),
  sessionUpload: effectiveFlag(wire, "session_upload"),
});

export const createSettingsApi = (http: HttpClient) => ({
  async get(): Promise<SettingsSnapshot | null> {
    const data = await http.get<Wire>("/settings");
    return data && parse(data);
  },

  async update(patch: SettingsPatch): Promise<SettingsSnapshot | null> {
    const data = await http.post<Wire>("/settings", patch);
    if (data) serverDefaults.bump();
    return data && parse(data);
  },

  async reset(): Promise<SettingsSnapshot | null> {
    const data = await http.post<Wire>("/settings/reset");
    if (data) serverDefaults.bump();
    return data && parse(data);
  },
});

export const settingsApi = createSettingsApi(http);

File: client/app/src/commonMain/kotlin/com/jahirtrap/cconnect/data/remote/SettingsApi.kt
package com.jahirtrap.cconnect.data.remote

import com.jahirtrap.cconnect.data.ServerDefaults
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put

// Backend-owned generation settings (the source of truth lives in the backend DB).
// CLI source/version is handled separately by CliApi.
object SettingsApi {

    data class Snapshot(
        val account: String,
        val model: String,
        val effort: String,
        val permissionMode: String,
        val streaming: Boolean,
        val todoTools: Boolean,
        val outputStyle: String,
        val mcpDisabled: String,
        val showThinking: String,
        val showToolUse: String,
        val showFileChange: String,
        val showCompact: String,
        val showWorking: String,
        val showTokens: Boolean,
        val simpleMode: Boolean,
        val chatOrder: String,
        val trashEnabled: Boolean,
        val defaultCategory: String,
        val retentionDays: Long,
        val retentionMin: Long,
        val retentionMax: Long,
    )

    private fun effectiveStr(o: JsonObject, key: String, fallback: String): String =
        o[key]?.jsonObject?.get("effective")?.jsonPrimitive?.contentOrNull ?: fallback

    private fun effectiveBool(o: JsonObject, key: String, fallback: Boolean): Boolean =
        o[key]?.jsonObject?.get("effective")?.jsonPrimitive?.booleanOrNull ?: fallback

    private fun effectiveInt(o: JsonObject, key: String, fallback: Int): Int =
        o[key]?.jsonObject?.get("effective")?.jsonPrimitive?.intOrNull ?: fallback

    private fun effectiveLong(o: JsonObject, key: String, fallback: Long): Long =
        o[key]?.jsonObject?.get("effective")?.jsonPrimitive?.longOrNull ?: fallback

    private fun boundLong(o: JsonObject, key: String, bound: String, fallback: Long): Long =
        o[key]?.jsonObject?.get(bound)?.jsonPrimitive?.longOrNull ?: fallback

    private fun parse(o: JsonObject) = Snapshot(
        account = effectiveStr(o, "account", ""),
        model = effectiveStr(o, "model", "opus"),
        effort = effectiveStr(o, "effort", "xhigh"),
        permissionMode = effectiveStr(o, "permission_mode", "bypassPermissions"),
        streaming = effectiveBool(o, "streaming", true),
        todoTools = effectiveBool(o, "todo_tools", false),
        outputStyle = effectiveStr(o, "output_style", "default"),
        mcpDisabled = effectiveStr(o, "mcp_disabled", ""),
        showThinking = effectiveStr(o, "show_thinking", "full"),
        showToolUse = effectiveStr(o, "show_tool_use", "label"),
        showFileChange = effectiveStr(o, "show_file_change", "full"),
        showCompact = effectiveStr(o, "show_compact", "full"),
        showWorking = effectiveStr(o, "show_working", "label"),
        showTokens = effectiveBool(o, "show_tokens", false),
        simpleMode = effectiveBool(o, "simple_mode", false),
        chatOrder = effectiveStr(o, "chat_order", "auto"),
        trashEnabled = effectiveBool(o, "trash_enabled", false),
        defaultCategory = effectiveStr(o, "default_category", ""),
        retentionDays = effectiveLong(o, "retention_days", 30),
        retentionMin = boundLong(o, "retention_days", "minimum", 1),
        retentionMax = boundLong(o, "retention_days", "maximum", Long.MAX_VALUE),
    )

    suspend fun get(): Snapshot? = Http.get("/settings")?.jsonObject?.let(::parse)

    suspend fun update(
        account: String? = null,
        model: String? = null,
        effort: String? = null,
        permissionMode: String? = null,
        streaming: Boolean? = null,
        todoTools: Boolean? = null,
        outputStyle: String? = null,
        mcpDisabled: String? = null,
        showThinking: String? = null,
        showToolUse: String? = null,
        showFileChange: String? = null,
        showCompact: String? = null,
        showWorking: String? = null,
        showTokens: Boolean? = null,
        simpleMode: Boolean? = null,
        chatOrder: String? = null,
        trashEnabled: Boolean? = null,
        defaultCategory: String? = null,
        retentionDays: Long? = null,
    ): Snapshot? = Http.post("/settings", buildJsonObject {
        if (account != null) put("account", account)
        if (model != null) put("model", model)
        if (effort != null) put("effort", effort)
        if (permissionMode != null) put("permission_mode", permissionMode)
        if (streaming != null) put("streaming", streaming)
        if (todoTools != null) put("todo_tools", todoTools)
        if (outputStyle != null) put("output_style", outputStyle)
        if (mcpDisabled != null) put("mcp_disabled", mcpDisabled)
        if (showThinking != null) put("show_thinking", showThinking)
        if (showToolUse != null) put("show_tool_use", showToolUse)
        if (showFileChange != null) put("show_file_change", showFileChange)
        if (showCompact != null) put("show_compact", showCompact)
        if (showWorking != null) put("show_working", showWorking)
        if (showTokens != null) put("show_tokens", showTokens)
        if (simpleMode != null) put("simple_mode", simpleMode)
        if (chatOrder != null) put("chat_order", chatOrder)
        if (trashEnabled != null) put("trash_enabled", trashEnabled)
        if (defaultCategory != null) put("default_category", defaultCategory)
        if (retentionDays != null) put("retention_days", retentionDays)
    })?.jsonObject?.let(::parse)?.also { ServerDefaults.bump() }

    suspend fun reset(): Snapshot? =
        Http.post("/settings/reset")?.jsonObject?.let(::parse)?.also { ServerDefaults.bump() }
}
